File: scripts/deploy.mjs
#!/usr/bin/env node
/**
 * deploy.mjs — Clean build + deploy to production
 *
 * Usage:
 *   npm run deploy
 *   node scripts/deploy.mjs
 */

import { execSync }   from 'child_process';
import { resolve, dirname } from 'path';
import { fileURLToPath }    from 'url';
import { rmSync, existsSync } from 'fs';
import { submitIndexNow } from './indexnow.mjs';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT      = resolve(__dirname, '..');

// ── Deployment configuration (environment-driven, public-key only) ───────────
//
// Required environment variables (deployment fails closed if any is missing):
//   CBW_DEPLOY_HOST        production host name or IP
//   CBW_DEPLOY_USER        deployment SSH user
//   CBW_DEPLOY_KEY_PATH    path to the private key used for public-key auth
//
// Optional environment variables (documented defaults):
//   CBW_DEPLOY_PORT        SSH port          (default: 22)
//   CBW_DEPLOY_REMOTE_PATH remote web root   (default: /var/www/cryptobonusworld/html)
//
// No password is read, accepted or supported. Host-key verification and
// public-key-only authentication are enforced on every ssh/scp call. See
// docs/security/CBW_DEPLOYMENT_CREDENTIALS_STANDARD_v1.md.

function failConfig(message) {
  console.error(`\nDeploy configuration error: ${message}`);
  process.exit(2);
}

function loadDeployConfig() {
  const host    = (process.env.CBW_DEPLOY_HOST     || '').trim();
  const user    = (process.env.CBW_DEPLOY_USER     || '').trim();
  const keyPath = (process.env.CBW_DEPLOY_KEY_PATH || '').trim();

  const missing = [
    ['CBW_DEPLOY_HOST', host],
    ['CBW_DEPLOY_USER', user],
    ['CBW_DEPLOY_KEY_PATH', keyPath],
  ].filter(([, v]) => !v).map(([name]) => name);

  if (missing.length) {
    failConfig(
      `missing required environment variable(s): ${missing.join(', ')}. ` +
      `Set them to production values in your local secret manager; do not ` +
      `hardcode them. Public-key authentication only.`
    );
  }

  const portRaw = (process.env.CBW_DEPLOY_PORT || '22').trim();
  const port = Number.parseInt(portRaw, 10);
  if (!Number.isInteger(port) || port <= 0) {
    failConfig('CBW_DEPLOY_PORT must be a positive integer.');
  }

  const remotePath =
    (process.env.CBW_DEPLOY_REMOTE_PATH || '').trim() ||
    '/var/www/cryptobonusworld/html';

  if (!existsSync(keyPath)) {
    failConfig(
      'CBW_DEPLOY_KEY_PATH does not point to a readable private key file. ' +
      'Provide the path to your deployment SSH private key.'
    );
  }

  return { host, user, keyPath, port, remotePath };
}

const CFG = loadDeployConfig();
const SERVER = `${CFG.user}@${CFG.host}`;
const WEB_ROOT = CFG.remotePath;

// Public-key-only, host-key-verified transport options for ssh and scp.
// BatchMode disables all interactive prompts; the client falls back to no
// other auth method, so a broken key fails closed instead of prompting.
const SSH_HARDENING = [
  '-o BatchMode=yes',
  '-o PasswordAuthentication=no',
  '-o KbdInteractiveAuthentication=no',
  '-o PreferredAuthentications=publickey',
  '-o IdentitiesOnly=yes',
  '-o StrictHostKeyChecking=yes',
].join(' ');

// scp uses -P for port; ssh uses -p. Key path is quoted for paths with spaces.
const SCP_BASE = `scp ${SSH_HARDENING} -P ${CFG.port} -i "${CFG.keyPath}"`;
const SSH_BASE = `ssh ${SSH_HARDENING} -p ${CFG.port} -i "${CFG.keyPath}"`;

function run(cmd, opts = {}) {
  console.log(`\n$ ${cmd}`);
  execSync(cmd, { stdio: 'inherit', cwd: ROOT, ...opts });
}

console.log('\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
console.log('  CryptoBonusWorld — Deploy');
console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

// 1. Clear dist + Astro cache
console.log('\n🗑️  Clearing dist and cache...');
if (existsSync(`${ROOT}/dist`)) {
  rmSync(`${ROOT}/dist`, { recursive: true, force: true });
  console.log('   ✓ dist removed');
}
if (existsSync(`${ROOT}/.astro`)) {
  rmSync(`${ROOT}/.astro`, { recursive: true, force: true });
  console.log('   ✓ .astro cache removed');
}

// 1b. Regenerate OG images (only if --regen-og flag is passed)
if (process.argv.includes('--regen-og')) {
  console.log('\n🖼️  Regenerating OG images...');
  run('python scripts/generate_og_images.py');
}

// 2. Build
console.log('\n🔨  Building...');
run('npx astro build');

// 3. Package
console.log('\n📦  Packaging...');
run('tar czf dist.tar.gz -C dist .');

// 4. Upload
console.log('\n⬆️   Uploading...');
run(`${SCP_BASE} dist.tar.gz ${SERVER}:/tmp/`);

// 5. Deploy on server
console.log('\n🚀  Deploying on server...');
run(`${SSH_BASE} ${SERVER} "rm -rf ${WEB_ROOT}/* && tar xzf /tmp/dist.tar.gz -C ${WEB_ROOT}/ && rm /tmp/dist.tar.gz && echo SERVER_DONE"`);

// 6. Cleanup local
if (existsSync(`${ROOT}/dist.tar.gz`)) {
  rmSync(`${ROOT}/dist.tar.gz`);
  console.log('\n🧹  Cleaned up dist.tar.gz');
}

console.log('\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
console.log('  ✅  Deploy complete!');
console.log('  🌐  https://cryptobonusworld.com');
console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n');

// 7. IndexNow — notify search engines of updated content
// Skip if --no-indexnow flag is passed (useful for debug/rollback deploys)
if (!process.argv.includes('--no-indexnow')) {
  console.log('🔔  Notifying search engines via IndexNow…');
  try {
    // Default deploy submits priority mode = tier1 money pages only
    // (/, /promo-codes/, /exchanges/, six live review pages).
    // Pass --full-indexnow for `all` mode = the full current sitemap set
    // (tier1 + trust/legal pages). Retired 301'd URLs are never submitted
    // on deploy — use `node scripts/indexnow.mjs --mode legacy` manually.
    const { ok, results } = await submitIndexNow(null, {
      silent: false,
      dryRun: process.argv.includes('--dry-run'),
      mode: process.argv.includes('--full-indexnow') ? 'all' : 'priority',
    });
    if (!ok) {
      console.warn('⚠️  IndexNow: some submissions failed — check logs/indexnow-submissions.jsonl');
    }
  } catch (e) {
    // Non-fatal — deploy succeeded even if IndexNow fails
    console.warn(`⚠️  IndexNow submission error (non-fatal): ${e.message}`);
  }
}

File: deploy.py
import paramiko
import os
import sys
import time

# ── Deployment configuration (environment-driven, public-key only) ────────────
#
# Required environment variables (deployment fails closed if any is missing):
#   CBW_DEPLOY_HOST        production host name or IP
#   CBW_DEPLOY_USER        deployment SSH user
#   CBW_DEPLOY_KEY_PATH    path to the private key used for public-key auth
#
# Optional environment variables (documented defaults):
#   CBW_DEPLOY_PORT        SSH port          (default: 22)
#   CBW_DEPLOY_REMOTE_PATH remote web root   (default: /var/www/cryptobonusworld/html)
#
# No password is read, accepted or supported. See
# docs/security/CBW_DEPLOYMENT_CREDENTIALS_STANDARD_v1.md.

DEFAULT_PORT = 22
DEFAULT_REMOTE_PATH = '/var/www/cryptobonusworld/html'

MAX_RETRIES = 3


def _fail(message):
    """Emit an actionable, non-sensitive configuration error and exit closed."""
    sys.stderr.write(f"Deploy configuration error: {message}\n")
    sys.stderr.flush()
    sys.exit(2)


def load_config():
    """Read and validate deployment configuration from the environment.

    Fails closed before any network operation. Never reads or accepts a
    password, a default host, a default user or a machine-specific key path.
    """
    host = os.environ.get('CBW_DEPLOY_HOST', '').strip()
    user = os.environ.get('CBW_DEPLOY_USER', '').strip()
    key_path = os.environ.get('CBW_DEPLOY_KEY_PATH', '').strip()

    missing = [name for name, value in (
        ('CBW_DEPLOY_HOST', host),
        ('CBW_DEPLOY_USER', user),
        ('CBW_DEPLOY_KEY_PATH', key_path),
    ) if not value]
    if missing:
        _fail(
            'missing required environment variable(s): '
            + ', '.join(missing)
            + '. Set them to production values in your local secret manager; '
            'do not hardcode them. Public-key authentication only.'
        )

    port_raw = os.environ.get('CBW_DEPLOY_PORT', str(DEFAULT_PORT)).strip()
    try:
        port = int(port_raw)
    except ValueError:
        _fail("CBW_DEPLOY_PORT must be an integer.")

    remote_root = os.environ.get('CBW_DEPLOY_REMOTE_PATH', DEFAULT_REMOTE_PATH).strip() \
        or DEFAULT_REMOTE_PATH

    key_path = os.path.expanduser(key_path)
    if not os.path.isfile(key_path):
        _fail(
            "CBW_DEPLOY_KEY_PATH does not point to a readable private key file. "
            "Provide the path to your deployment SSH private key."
        )

    local_dist = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'dist')
    if not os.path.isdir(local_dist):
        _fail("local dist/ directory not found — build the site before deploying.")

    return {
        'host': host,
        'user': user,
        'port': port,
        'key_path': key_path,
        'remote_root': remote_root,
        'local_dist': local_dist,
    }


def make_client(cfg):
    """Open a strict, public-key-only SSH connection.

    Host keys must already be known (system/user known_hosts); unknown hosts
    are rejected. Agent and on-disk key discovery are disabled so only the
    explicitly supplied key is used. No password is ever passed.
    """
    client = paramiko.SSHClient()
    client.load_system_host_keys()
    client.set_missing_host_key_policy(paramiko.RejectPolicy())
    try:
        client.connect(
            cfg['host'],
            port=cfg['port'],
            username=cfg['user'],
            key_filename=cfg['key_path'],
            allow_agent=False,
            look_for_keys=False,
            timeout=30,
            banner_timeout=30,
            auth_timeout=30,
        )
    except paramiko.ssh_exception.SSHException as exc:
        # Message may reference host-key rejection or auth failure, never a secret.
        _fail(f"SSH connection failed ({exc.__class__.__name__}): {exc}")
    transport = client.get_transport()
    transport.set_keepalive(15)          # send keepalive every 15 s
    return client


def upload_file_with_retry(sftp, local_path, remote_path, retries=MAX_RETRIES):
    for attempt in range(retries):
        try:
            sftp.put(local_path, remote_path)
            return
        except Exception:
            if attempt < retries - 1:
                sys.stdout.write(f"  retry {attempt+1}: {os.path.basename(local_path)}\n")
                sys.stdout.flush()
                time.sleep(2)
            else:
                raise


def upload_dir(sftp, local_dir, remote_dir):
    try:
        sftp.stat(remote_dir)
    except FileNotFoundError:
        sftp.mkdir(remote_dir)

    for item in sorted(os.listdir(local_dir)):
        local_path = os.path.join(local_dir, item)
        remote_path = remote_dir + '/' + item
        if os.path.isdir(local_path):
            upload_dir(sftp, local_path, remote_path)
        else:
            upload_file_with_retry(sftp, local_path, remote_path)


# ── Main ────────────────────────────────────────────────────────────────────

def main():
    cfg = load_config()

    client = make_client(cfg)
    sys.stdout.write("Connected. Clearing remote html dir...\n")
    sys.stdout.flush()

    remote_root = cfg['remote_root']
    stdin, stdout, stderr = client.exec_command(f'rm -rf {remote_root}/* && echo done')
    result = stdout.read().strip()
    sys.stdout.write(f"Cleared: {result.decode()}\n")
    sys.stdout.flush()

    sftp = client.open_sftp()
    sys.stdout.write("Uploading dist/...\n")
    sys.stdout.flush()

    try:
        upload_dir(sftp, cfg['local_dist'], remote_root)
        sftp.close()
        client.close()
        sys.stdout.write("Deploy complete.\n")
        sys.stdout.flush()
    except Exception as exc:
        sys.stdout.write(f"Deploy failed: {exc}\n")
        sys.stdout.flush()
        sftp.close()
        client.close()
        sys.exit(1)


if __name__ == '__main__':
    main()
